hypothesis-python/src/hypothesis/internal/intervalsets.py

```diff
@@ -8,12 +8,17 @@
 # v. 2.0. If a copy of the MPL was not distributed with this file, You can
 # obtain one at https://mozilla.org/MPL/2.0/.
 
-from typing import Union
+from typing import TYPE_CHECKING, Iterable, List, Sequence, Tuple, Union, cast
+
+if TYPE_CHECKING:
+    from typing import TypeAlias
+
+IntervalsT: "TypeAlias" = Tuple[Tuple[int, int], ...]
 
 
 class IntervalSet:
     @classmethod
-    def from_string(cls, s):
+    def from_string(cls, s: str) -> "IntervalSet":
```

Comment:
1. I didn't because our mypy is on 3.10 and `Self` was added in 3.11 – but, as I just learned, mypy depends on `typing_extensions` and so we can import `Self` from there
2. but apparently, strictly speaking, `Self` is not correct because of:

> You should not use Self as the return annotation if the method is not guaranteed to return an instance of a subclass when the class is subclassed
> https://docs.python.org/3/library/typing.html#typing.Self

I'll rejib `{union, difference, ...}` to be classmethods returning `cls(...)`, which is probably cleaner anyway if we ever do subclass.

Reply:
I'm not sure this is any better actually. The `x.union(y)` syntax is nicer than `IntervalSet.union(x, y)` (though `x | y` is better yet). Maybe we just settle for not using `Self`? I'm ambivalent – curious how bad you think `IntervalSet.union(x, y)` is.

Reply:
I'm fine with using `Self` and fibbing a bit about the details, but we could also do `return type(self)(...)` instead of `IntervalSet(...)` in a few methods to make it true.

Reply:
oh awesome, `@final` works beautifully!